WBCore/GeolocationHelper.swift

```diff
@@ -0,0 +1,188 @@
+    }
+
+    func onAuthorizationStatusIsGranted() {
+        locationManager.startUpdatingLocation();
+    }
+
+    func onAuthorizationStatusIsDenied() {
+        webView.evaluateJavaScript("navigator.geolocation.helper.error(1, 'App does not have location permission');");
+    }
+
+    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
+        if (message.name == "listenerAdded") {
+            listenersCount += 1;
+
+            if (!locationServicesIsEnabled()) {
+                onLocationServicesIsDisabled();
+            }
+            else if (authorizationStatusIsDenied(status: CLLocationManager.authorizationStatus())) {
+                onAuthorizationStatusIsDenied();
+            }
+            else if (authorizationStatusNeedRequest(status: CLLocationManager.authorizationStatus())) {
+                onAuthorizationStatusNeedRequest();
+            }
+            else if (authorizationStatusIsGranted(status: CLLocationManager.authorizationStatus())) {
+                onAuthorizationStatusIsGranted();
+            }
+        }
+        else if (message.name == "listenerRemoved") {
+            listenersCount -= 1;
+
+            // no listener left in web view to wait for position
+            if (listenersCount == 0) {
+                locationManager.stopUpdatingLocation();
+            }
+        }
+    }
+
+    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
+        // didChangeAuthorization is also called at app startup, so this condition checks listeners
+        // count before doing anything otherwise app will start location service without reason
+        if (listenersCount > 0) {
+            if (authorizationStatusIsDenied(status: status)) {
+                onAuthorizationStatusIsDenied();
+            }
+            else if (authorizationStatusIsGranted(status: status)) {
+                onAuthorizationStatusIsGranted();
+            }
+        }
+    }
+
+    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
+        if let location = locations.last {
+            webView.evaluateJavaScript("navigator.geolocation.helper.success('\(location.timestamp)', \(location.coordinate.latitude), \(location.coordinate.longitude), \(location.altitude), \(location.horizontalAccuracy), \(location.verticalAccuracy), \(location.course), \(location.speed));");
+        }
+    }
+
+    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
+        webView.evaluateJavaScript("navigator.geolocation.helper.error(2, 'Failed to get position (\(error.localizedDescription))');");
+    }
+
+    func getJavaScripToEvaluate() -> String {
+        let javaScripToEvaluate = """
+            // management for success and error listeners and its calling
+            navigator.geolocation.helper = {
```

Comment:
I'm currently on holiday so will review and try these changes out more fully later, but in the meantime, is this interface following a spec? If so, can you provide a link to it? Also, it would be nice if this javascript code could be moved to a bundle resource so that it can be edited as javascript rather than as a string within the swift code.

Reply:
Sure! https://developer.mozilla.org/en-US/docs/Web/API/Geolocation